File: src/claude/permission-handler.ts
import { Api } from "grammy";
import { buildPermissionKeyboard } from "../telegram/keyboard-builder.js";
import { logger } from "../utils/logger.js";

interface PendingPermission {
  resolve: (allowed: boolean) => void;
  toolName: string;
  telegramMessageId?: number;
}

/**
 * Manages permission requests from Claude SDK.
 * When Claude asks to use a tool, sends inline buttons to Telegram
 * and waits for user response via a deferred Promise.
 */
export class PermissionHandler {
  private pending = new Map<string, PendingPermission>();
  private counter = 0;
  private allowAllSession = false;

  constructor(
    private api: Api,
    private chatId: number,
  ) {}

  /** Generate a unique request ID */
  private nextId(): string {
    return `${this.chatId}:${++this.counter}`;
  }

  /**
   * Called by the SDK when a tool needs permission.
   * Returns a Promise that resolves when the user responds.
   */
  async requestPermission(
    toolName: string,
    toolInput: Record<string, unknown>,
  ): Promise<boolean> {
    // Auto-allow if user chose "Allow All" for this session
    if (this.allowAllSession) {
      logger.debug({ toolName }, "Auto-allowed (session allow all)");
      return true;
    }

    const requestId = this.nextId();

    // Format tool info for display
    let detail = "";
    if ("command" in toolInput) {
      detail = `\nCommand: ${String(toolInput["command"]).slice(0, 200)}`;
    } else if ("file_path" in toolInput) {
      detail = `\nFile: ${toolInput["file_path"]}`;
    } else if ("pattern" in toolInput) {
      detail = `\nPattern: ${toolInput["pattern"]}`;
    }

    const text = `🔒 Permission request\n\nTool: ${toolName}${detail}`;

    const keyboard = buildPermissionKeyboard(requestId);

    return new Promise<boolean>((resolve) => {
      // Send the permission request message
      void this.api
        .sendMessage(this.chatId, text, { reply_markup: keyboard })
        .then((msg) => {
          const pending = this.pending.get(requestId);
          if (pending) {
            pending.telegramMessageId = msg.message_id;
          }
        })
        .catch((error) => {
          logger.error({ error }, "Failed to send permission request");
          resolve(false);
        });

      this.pending.set(requestId, { resolve, toolName });

      // Timeout after 5 minutes
      setTimeout(() => {
        if (this.pending.has(requestId)) {
          this.pending.delete(requestId);
          resolve(false);
          logger.warn({ requestId, toolName }, "Permission request timed out");
        }
      }, 5 * 60 * 1000);
    });
  }

  /**
   * Handle callback from inline button press.
   * Returns true if this callback was handled.
   */
  async handleCallback(
    data: string,
    answerCallback: () => Promise<void>,
  ): Promise<boolean> {
    const parts = data.split(":");
    if (parts[0] !== "perm") return false;

    const action = parts[1];
    const requestId = parts.slice(2).join(":");
    const pending = this.pending.get(requestId);
    if (!pending) return false;

    this.pending.delete(requestId);

    let allowed = false;
    let statusText: string;

    switch (action) {
      case "allow":
        allowed = true;
        statusText = `✅ Allowed: ${pending.toolName}`;
        break;
      case "deny":
        allowed = false;
        statusText = `❌ Denied: ${pending.toolName}`;
        break;
      case "allow_all":
        allowed = true;
        this.allowAllSession = true;
        statusText = `✅ Allowed all for this session: ${pending.toolName}`;
        break;
      default:
        return false;
    }

    // Update the message to show the result
    if (pending.telegramMessageId) {
      try {
        await this.api.editMessageText(
          this.chatId,
          pending.telegramMessageId,
          statusText,
        );
      } catch {
        // Message may have been deleted
      }
    }

    await answerCallback();
    pending.resolve(allowed);
    return true;
  }

  /** Reset session-level auto-allow */
  resetAllowAll(): void {
    this.allowAllSession = false;
  }

  /** Clean up pending requests */
  cleanup(): void {
    for (const [, pending] of this.pending) {
      pending.resolve(false);
    }
    this.pending.clear();
  }
}

// Active permission handlers per chat
const handlers = new Map<number, PermissionHandler>();

export function getPermissionHandler(
  api: Api,
  chatId: number,
): PermissionHandler {
  let handler = handlers.get(chatId);
  if (!handler) {
    handler = new PermissionHandler(api, chatId);
    handlers.set(chatId, handler);
  }
  return handler;
}

export function findPermissionHandler(
  chatId: number,
): PermissionHandler | undefined {
  return handlers.get(chatId);
}

File: proxima/claude/permission_handler.py
from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from dataclasses import dataclass

from aiogram import Bot

from proxima.logging import get_logger
from proxima.telegram.keyboards import build_permission_keyboard

logger = get_logger(__name__)


@dataclass
class PendingPermission:
    future: asyncio.Future[bool]
    tool_name: str
    telegram_message_id: int | None = None


class PermissionHandler:
    def __init__(self, bot: Bot, chat_id: int, message_thread_id: int | None = None) -> None:
        self.bot = bot
        self.chat_id = chat_id
        self.message_thread_id = message_thread_id
        self._pending: dict[str, PendingPermission] = {}
        self._counter = 0
        self._allow_all_session = False

    def _next_id(self) -> str:
        self._counter += 1
        return f"{self.chat_id}:{self._counter}"

    async def request_permission(self, tool_name: str, tool_input: dict[str, object]) -> bool:
        if self._allow_all_session:
            logger.debug("permission_auto_allow", tool=tool_name)
            return True

        logger.info("permission_requested", tool=tool_name, chat_id=self.chat_id)
        request_id = self._next_id()
        detail = ""
        if "command" in tool_input:
            detail = f"\nCommand: {str(tool_input['command'])[:200]}"
        elif "file_path" in tool_input:
            detail = f"\nFile: {tool_input['file_path']}"
        elif "pattern" in tool_input:
            detail = f"\nPattern: {tool_input['pattern']}"

        text = f"Permission request\n\nTool: {tool_name}{detail}"
        keyboard = build_permission_keyboard(request_id)

        loop = asyncio.get_running_loop()
        future: asyncio.Future[bool] = loop.create_future()
        self._pending[request_id] = PendingPermission(future=future, tool_name=tool_name)

        try:
            message = await self.bot.send_message(
                chat_id=self.chat_id,
                text=text,
                reply_markup=keyboard,
                message_thread_id=self.message_thread_id,
            )
            self._pending[request_id].telegram_message_id = message.message_id
        except Exception as exc:  # noqa: BLE001
            logger.exception("permission_message_send_failed", error=str(exc))
            self._pending.pop(request_id, None)
            return False

        async def timeout() -> None:
            await asyncio.sleep(5 * 60)
            pending = self._pending.pop(request_id, None)
            if pending and not pending.future.done():
                pending.future.set_result(False)
                logger.warning("permission_request_timeout", request_id=request_id)

        asyncio.create_task(timeout())
        return await future

    async def handle_callback(
        self,
        data: str,
        answer_callback: Callable[[], Awaitable[None]],
    ) -> bool:
        parts = data.split(":")
        if not parts or parts[0] != "perm":
            return False

        if len(parts) < 4:
            return False

        action = parts[1]
        request_id = ":".join(parts[2:])
        pending = self._pending.pop(request_id, None)
        if pending is None:
            return False

        allowed = False
        status_text = "Denied"

        if action == "allow":
            allowed = True
            status_text = f"Allowed: {pending.tool_name}"
        elif action == "deny":
            allowed = False
            status_text = f"Denied: {pending.tool_name}"
        elif action == "allow_all":
            allowed = True
            self._allow_all_session = True
            status_text = f"Allowed all for session: {pending.tool_name}"
        else:
            return False

        logger.info(
            "permission_callback",
            action=action,
            tool=pending.tool_name,
            chat_id=self.chat_id,
        )

        if pending.telegram_message_id:
            try:
                await self.bot.edit_message_text(
                    chat_id=self.chat_id,
                    message_id=pending.telegram_message_id,
                    text=status_text,
                )
            except Exception:
                pass

        await answer_callback()
        if not pending.future.done():
            pending.future.set_result(allowed)
        return True

    def reset_allow_all(self) -> None:
        self._allow_all_session = False

    def cleanup(self) -> None:
        for pending in self._pending.values():
            if not pending.future.done():
                pending.future.set_result(False)
        self._pending.clear()


_handlers: dict[tuple[int, int | None], PermissionHandler] = {}


def get_permission_handler(
    bot: Bot, chat_id: int, thread_id: int | None = None
) -> PermissionHandler:
    key = (chat_id, thread_id)
    handler = _handlers.get(key)
    if not handler:
        handler = PermissionHandler(bot, chat_id, message_thread_id=thread_id)
        _handlers[key] = handler
    return handler


def find_permission_handler(
    chat_id: int, thread_id: int | None = None
) -> PermissionHandler | None:
    return _handlers.get((chat_id, thread_id))
